src/core/credentials/transport/xds/xds_credentials.cc

```diff
@@ -105,12 +138,20 @@ void XdsCertificateVerifier::Cancel(
 int XdsCertificateVerifier::CompareImpl(
     const grpc_tls_certificate_verifier* other) const {
   auto* o = static_cast<const XdsCertificateVerifier*>(other);
+  int compare_cert_provider;
   if (xds_certificate_provider_ == nullptr ||
       o->xds_certificate_provider_ == nullptr) {
-    return QsortCompare(xds_certificate_provider_,
-                        o->xds_certificate_provider_);
+    compare_cert_provider =
+        QsortCompare(xds_certificate_provider_, o->xds_certificate_provider_);
+  } else {
+    compare_cert_provider =
+        xds_certificate_provider_->Compare(o->xds_certificate_provider_.get());
+  }
+  if (compare_cert_provider == 0) {
```

Comment:
I think this logic would be a bit easier to follow if we reversed the conditional here:

```
if (compare_cert_provider != 0) return compare_cert_provider;
return sni_matcher_.string_matcher().compare(
    o->sni_matcher_.string_matcher());
```